src/main/java/aztech/modern_industrialization/machines/components/CustomOverclockComponent.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Azercoco & Technici4n
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package aztech.modern_industrialization.machines.components;
+
+import aztech.modern_industrialization.MIText;
+import aztech.modern_industrialization.MITooltips;
+import aztech.modern_industrialization.machines.MachineBlockEntity;
+import java.util.ArrayList;
+import java.util.List;
+import net.minecraft.core.Registry;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+
+public class CustomOverclockComponent implements IOverclockComponent {
```

Comment:
This should be merged with the gunpowder component as it's a generalization of it.

Reply:
It's certainly possible to make gunpowder just configuration in the CustomOverclockComponent since it contains more functionality. I only left it separate so there were less chance of breaking existing MI gunpowder behavior. From a maintainability standpoint it would be better to have one. If I remove the GunpowderOverClockComponent, should I remove the IOverclockComponent interface and rename CustomOverClockComponent to just OverclockComponent?

Reply:
Minor breaking change but GunpowderOverclockComponent writes "overclockGunpowderTick" NBT where as the new OverlockComponent writes "overclockTicks" NBT which will result in machines losing their gunpowder overclock ticks when upgrading to the new version. Is this an issue?

Reply:
Yes, the format should be kept compatible. In the reading method, you can check for both. :Wink:

Make sure to add a `// TODO 1.20: Remove upgrade code`.